fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java

```diff
@@ -370,4 +420,179 @@ public void updateRecords() {
+        assertNotNull(preUpgradeMetaData.getIndex("MySimpleRecord$num_value_3_indexed"));
+        assertNotNull(preUpgradeMetaData.getIndex("preUpgradeIndex"));
+        assertEquals(preUpgradeMetaData.getIndex("MySimpleRecord$str_value_indexed").getName(),
+                preUpgradeMetaData.getIndex("MySimpleRecord$str_value_indexed").getSubspaceKey());
+        assertEquals(preUpgradeMetaData.getIndex("MySimpleRecord$num_value_unique").getName(),
+                preUpgradeMetaData.getIndex("MySimpleRecord$num_value_unique").getSubspaceKey());
+        assertEquals(preUpgradeMetaData.getIndex("MySimpleRecord$num_value_3_indexed").getName(),
+                preUpgradeMetaData.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey());
+        assertEquals(preUpgradeMetaData.getIndex("preUpgradeIndex").getName(),
+                preUpgradeMetaData.getIndex("preUpgradeIndex").getSubspaceKey());
+
+        // User has chosen not to enable counter-based subspace keys after upgrade.
+        RecordMetaDataBuilder afterUpgradeBuilder = RecordMetaData.newBuilder().setRecords(preUpgradeMetaData.toProto());
+        afterUpgradeBuilder.addIndex("MySimpleRecord", "postUpgradeIndex", "num_value_2");
+        RecordMetaData afterUpgradeMetaData = afterUpgradeBuilder.build(true);
+        assertNotNull(afterUpgradeMetaData.getIndex("MySimpleRecord$str_value_indexed"));
+        assertNotNull(afterUpgradeMetaData.getIndex("MySimpleRecord$num_value_unique"));
+        assertNotNull(afterUpgradeMetaData.getIndex("MySimpleRecord$num_value_3_indexed"));
+        assertNotNull(afterUpgradeMetaData.getIndex("preUpgradeIndex"));
+        assertNotNull(afterUpgradeMetaData.getIndex("postUpgradeIndex"));
+        assertEquals(afterUpgradeMetaData.getIndex("MySimpleRecord$str_value_indexed").getName(),
+                afterUpgradeMetaData.getIndex("MySimpleRecord$str_value_indexed").getSubspaceKey());
+        assertEquals(afterUpgradeMetaData.getIndex("MySimpleRecord$num_value_unique").getName(),
+                afterUpgradeMetaData.getIndex("MySimpleRecord$num_value_unique").getSubspaceKey());
+        assertEquals(afterUpgradeMetaData.getIndex("MySimpleRecord$num_value_3_indexed").getName(),
+                afterUpgradeMetaData.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey());
+        assertEquals(afterUpgradeMetaData.getIndex("preUpgradeIndex").getName(),
+                afterUpgradeMetaData.getIndex("preUpgradeIndex").getSubspaceKey());
+        assertEquals(afterUpgradeMetaData.getIndex("postUpgradeIndex").getName(),
+                afterUpgradeMetaData.getIndex("postUpgradeIndex").getSubspaceKey());
+
+        // But later user decides to enable counter-based subspace keys assignment. Previous indexes will use name, newer indexes will use the counter.
+        RecordMetaDataBuilder builderWithCounter = RecordMetaData.newBuilder().enableCounterBasedSubspaceKeys().setRecords(afterUpgradeMetaData.toProto());
```

Comment:
So, this is *almost* the test I want. I the problem is that the indexes in `afterUpgradeMetaData.toProto` should all have subspace keys set within the proto (because of the change that was made in this PR to always serialize it). The path left untested here is a meta-data proto where an index does *not* have the `subspace_key` set but counter-based assignment is enabled. (In other words, the thing that will happen when any meta-data store user enables this feature for the first time.)

Reply:
Ah, great, yes, that's what I mean!